File: R/cache.R
#' Where tidyeavs stores downloaded data
#'
#' EAVS data files are downloaded once and kept in a local cache, so that
#' later calls reuse the file instead of fetching it again. This returns the
#' cache directory.
#'
#' The location is resolved in this order:
#'
#' 1. the `tidyeavs.cache_dir` option, if set;
#' 2. the `TIDYEAVS_CACHE_DIR` environment variable, if set;
#' 3. `tools::R_user_dir("tidyeavs", "cache")`, the standard per-user cache
#'    directory for R packages.
#'
#' To use a project-local cache for one session, set the option before
#' downloading, e.g. `options(tidyeavs.cache_dir = "data/eavs")`.
#'
#' @param create Whether to create the directory if it does not yet exist.
#'   Defaults to `TRUE`.
#'
#' @return The cache directory path, as a string.
#' @export
#'
#' @examples
#' # Where files will be stored (without creating the directory):
#' eavs_cache_dir(create = FALSE)
eavs_cache_dir <- function(create = TRUE) {
  dir <- getOption("tidyeavs.cache_dir", default = NULL)
  if (is.null(dir)) {
    dir <- Sys.getenv("TIDYEAVS_CACHE_DIR", unset = "")
    if (!nzchar(dir)) {
      dir <- tools::R_user_dir("tidyeavs", "cache")
    }
  }
  if (create && !dir.exists(dir)) {
    dir.create(dir, recursive = TRUE, showWarnings = FALSE)
  }
  dir
}

#' List files in the tidyeavs cache
#'
#' @return A tibble with one row per cached file: its `file` name, size in
#'   bytes (`bytes`), and last-modified time (`modified`). Zero rows if the
#'   cache is empty.
#' @export
eavs_cache_list <- function() {
  dir <- eavs_cache_dir(create = FALSE)
  if (!dir.exists(dir)) {
    return(tibble::tibble(
      file = character(), bytes = numeric(),
      modified = as.POSIXct(character())
    ))
  }
  files <- list.files(dir, full.names = TRUE, recursive = TRUE)
  info <- file.info(files)
  tibble::tibble(
    file = sub(paste0("^", dir, .Platform$file.sep), "", files),
    bytes = info$size,
    modified = info$mtime
  )
}

#' Delete files from the tidyeavs cache
#'
#' @param files Names of files to delete, as returned by `eavs_cache_list()`.
#'   If `NULL` (the default), every file in the cache is removed.
#' @param confirm Whether to ask before deleting when running interactively.
#'   Defaults to `TRUE`.
#'
#' @return The names of the deleted files, invisibly.
#' @export
eavs_cache_clear <- function(files = NULL, confirm = TRUE) {
  dir <- eavs_cache_dir(create = FALSE)
  if (!dir.exists(dir)) {
    return(invisible(character()))
  }
  if (is.null(files)) {
    files <- eavs_cache_list()$file
  }
  if (length(files) == 0) {
    return(invisible(character()))
  }
  if (confirm && interactive()) {
    msg <- sprintf(
      "Delete %d file%s from the tidyeavs cache?",
      length(files), if (length(files) == 1) "" else "s"
    )
    if (!isTRUE(utils::askYesNo(msg))) {
      return(invisible(character()))
    }
  }
  unlink(file.path(dir, files), recursive = TRUE)
  invisible(files)
}

# Full path to a named file in the cache (does not check existence).
cache_path <- function(file_name) {
  file.path(eavs_cache_dir(create = TRUE), file_name)
}

File: py/src/tidyeavs/cache.py
"""Where tidyeavs stores downloaded data."""

from __future__ import annotations

import os
import shutil
import sys
from pathlib import Path

import pandas as pd

_override: Path | None = None


def set_cache_dir(path: str | os.PathLike[str] | None) -> None:
    """Set the cache directory for this session.

    The Python analogue of R's ``options(tidyeavs.cache_dir = ...)``. Pass
    ``None`` to restore the default lookup.
    """
    global _override
    _override = None if path is None else Path(path)


def cache_dir(create: bool = True) -> Path:
    """Return the directory downloaded EAVS files are kept in.

    Resolved in this order: an explicit :func:`set_cache_dir`, the
    ``TIDYEAVS_CACHE_DIR`` environment variable, then the per-user cache
    directory for the platform.

    The R package honours ``TIDYEAVS_CACHE_DIR`` too, so setting it in both
    gives one shared cache and a file downloaded from either language is reused
    by the other. The platform defaults deliberately differ, since R resolves
    its own via ``tools::R_user_dir()``.
    """
    if _override is not None:
        path = _override
    else:
        env = os.environ.get("TIDYEAVS_CACHE_DIR", "")
        path = Path(env) if env else _platform_cache_dir()

    if create:
        path.mkdir(parents=True, exist_ok=True)
    return path


def _platform_cache_dir() -> Path:
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Caches" / "tidyeavs"
    if os.name == "nt":
        base = os.environ.get("LOCALAPPDATA") or (Path.home() / "AppData" / "Local")
        return Path(base) / "tidyeavs" / "Cache"
    xdg = os.environ.get("XDG_CACHE_HOME", "")
    return (Path(xdg) if xdg else Path.home() / ".cache") / "tidyeavs"


def cache_list() -> pd.DataFrame:
    """List cached files, with one row per file.

    Columns are ``file`` (the name, relative to the cache directory), ``bytes``,
    and ``modified``. Zero rows if the cache is empty or absent.
    """
    root = cache_dir(create=False)
    rows = []
    if root.is_dir():
        for path in sorted(root.rglob("*")):
            if path.is_file():
                stat = path.stat()
                rows.append(
                    {
                        "file": str(path.relative_to(root)),
                        "bytes": stat.st_size,
                        "modified": pd.Timestamp(stat.st_mtime, unit="s"),
                    }
                )
    return pd.DataFrame(rows, columns=["file", "bytes", "modified"])


def cache_clear(files: list[str] | None = None) -> list[str]:
    """Delete files from the cache and return the names removed.

    ``files`` names entries as :func:`cache_list` reports them; ``None`` (the
    default) removes everything. Unlike the R function there is no interactive
    confirmation, since Python callers are usually scripts. Nothing outside the
    cache directory is touched.
    """
    root = cache_dir(create=False)
    if not root.is_dir():
        return []

    names = list(cache_list()["file"]) if files is None else list(files)
    removed = []
    for name in names:
        target = (root / name).resolve()
        # Refuse a name that escapes the cache directory.
        if root.resolve() not in target.parents:
            raise ValueError(f"{name!r} is outside the cache directory")
        if target.is_dir():
            shutil.rmtree(target)
            removed.append(name)
        elif target.exists():
            target.unlink()
            removed.append(name)
    return removed


def cache_path(file_name: str) -> Path:
    """Full path to a named file in the cache. Does not check existence."""
    return cache_dir(create=True) / file_name
